forum/qa-include/pages/user-profile.php

```diff
@@ -270,7 +270,11 @@
 
     }
 
-
+    if (qa_clicked('doblockpw')) {

```

Comment:
ogolnie to nie jestem pewien, ale wydaje mi się, że nie powinno się nadpisywać qa-include/pages/* skoro mechanizm tworzenia pluginów takie coś umożliwia: https://docs.question2answer.org/plugins/layers/

Reply:
https://docs.question2answer.org/plugins/modules-page/

Reply:
Rozważaliśmy wprowadzenie tego ficzeru jako wtyczka, ale mechanizm pluginów nie oferował jakiejś istotnej funkcji do rozwiązania tego problemu, więc mieliśmy dwa wyjścia:
-albo zrobić połowę kodu w pluginie a połowę w silniku bezpośrednio
-albo wszystko robić w silniku
Wybrałem to drugie, nie wyobrażam sobie żeby część była w pluginie a część nie.

Reply:
nie widzę w historii rozmowy tych rozważań - nie wiem co to za funkcja, ale jestem niemal pewien, że da się to zrobić bez ingerencji w core

Reply:
No tutaj tego nie ma, rozmawialiśmy chyba gdzieś na IRC jeszcze o tym, generalnie większość tego kodu da się przenieść do wtyczki, problem pojawia się z tego co pamiętam przy przechwytywaniu wysyłania wiadomości oraz renderowaniu custom boxa w `message.php`

Reply:
@awaluk odniesiesz się? 
Moim zdaniem każdy kawałek kodu można zrobić bez ingerencji w core q2a. 
Change my mind

Reply:
Ciężko mi się odnieść. Czy była taka rozmowa i takie próby to nie pamiętam, zbyt długo to nie było ruszane. Czy każdy kawałek kodu można nadpisać też nie wiem - nie zaprzeczę i nie potwierdzę. Myślę że większość się da, ale czy na pewno _każdy_ to się nie odważę powiedzieć. Być może też nawet jeśli się da to może być to znacznie bardziej trudne i czasochłonne niż zmienić bezpośrednio. Jeśli powiedzmy to zwiększa czas pracy nad zmianą dwukrotnie to ok, ale jak zwiększa pięciokrotnie to już nie wiem czy na pewno ma to sens.

W związku z tym zgadzam się jak najbardziej, że najlepiej, aby zmiany były na pluginach, fajnie jakbyśmy tak próbowali, ale nie mogę ze 100% pewnością powiedzieć, że to się zawsze uda i będzie najlepszym rozwiązaniem. Zgodzić się też wypada z Mariuszem, że jeśli plugin ma być niekompletny to niech już lepiej wcale nie będzie pluginem, bo ktoś go kiedyś go wyciągnie z zamysłem że to normalny plugin i się posypie. Jeśli uda się całość przenieść na plugin to jak najbardziej fajnie.

Reply:
> Jeśli powiedzmy to zwiększa czas pracy nad zmianą dwukrotnie to ok, ale jak zwiększa pięciokrotnie to już nie wiem czy na pewno ma to sens.
Oczywiście, że ma sens. Nie ma tu jiry, żeby sobie trackować czas. Jest pełno PR, które wiszą tu kilka lat i nikt nie ma z tym problemu.

Reply:
Jak dla mnie nie ma. Nikt nie ma problemu, bo nikt nie wymaga od nikogo pracy nad daną rzeczą. Jak ktoś coś robił, kolokwialnie mówiąc mu się odechciało i nie robi no to ok, rozumiem to, nie płacimy za pracę, nie ma obowiązku. Ale jeśli już ktoś chce coś dla nas robić i zapyta mnie, to niech robi coś co ma sens.

Przykładowo: jak ktoś ma spędzić nad zrobieniem 1 rzeczy jako plugin tydzień, albo w tym czasie może zrobić tę 1 rzecz bez plugina w dzień i do tego zrobić jeszcze 5 innych rzeczy w pozostałe dni tygodnia, to wybrałbym wariant pierwszy. Oczywiście zależy od sytuacji, złożoności problemu, konkretnych potrzeb. Super byłoby aby ten ktoś zrobił plugin, ale jak w tym czasie nadal zrobi to dobrze tylko że bez pluginu i dostarczy nam 5 innych potrzebnych rzeczy no to sorry, ale jednak zysk funkcjonalny jest znacznie większy i zyskają użytkownicy na co dzień korzystający z forum. Też bym chciał żeby wszystko było idealnie, ale nie zawsze to jest najlepsze wyjście gdy ma sie ograniczone zasoby i czas.

Reply:
> Przykładowo: jak ktoś ma spędzić nad zrobieniem 1 rzeczy jako plugin tydzień, albo w tym czasie może zrobić tę 1 rzecz bez plugina w dzień i do tego zrobić jeszcze 5 innych rzeczy w pozostałe dni tygodnia, to wybrałbym wariant pierwszy.

Aż nie mogę wyjść z podziwu, że takie masz zdanie mając na uwadze fakt, że od ponad 30 godzin walczę nad jednym plikiem który był zrobiony na kolanie BEZ PLUGINU bo było szybciej. 
Nie po to twórca q2a stworzył możliwość tworzenia praktycznie wszystkiego niezależnie bez ingerencji w kod projektu żeby teraz wrzucać tam co się chce. Przez takie akcje aktualizacja wersji forum - a co za tym idzie, przyspieszenie jego działania, zwiększenie bezpieczeństwa i nowe funkcjonalności które są dostępne bez pluginów - staje się na tyle uciążliwa, że niemal niemożliwa. 

Serio wygoda i szybkość wdrożenia zmiany mają być jedynymi argumentami w tej dyskusji? Nie ma przymusu - skoro ktoś zaczął coś robić, to niech chociaż robi tak, żeby nie utrudniać innym pracy. Po co mi milion funkcjonalności w kodzie, które są naszyte w całym core, skoro później muszę zapewne miesiąc jak i nie dłużej walczyć z upgrade wersji forum? Skoro są to dodatkowe funkcjonalności to nie są "must have" a skoro nie są must have, to niech chociaż będą przyjazne do wdrożenia i rozwijania - nawet kosztem czasu. Masa rzeczy tu wisi latami i dalej nie jestem przekonany do Twojej argumentacji. 
Sam wiesz, że nic tu się nie zyskuje słabymi rozwiązaniami, a otrzymuje się w zamian straszny bagaż.

Reply:
Nigdzie nie napisałem, że szybkość i wygoda to jedyne argumenty. Przeciwnie - kilkukrotnie zaznaczyłem, że najlepiej by było, aby było to zrobione jako plugin i należałoby do tego dążyć. Jednak również ważny jest czas i możliwość wprowadzenia zmian, a Ty zdaje się, że jako jedyny argument widzisz kod. Trzeba to wypośrodkować - robić możliwie najlepszy kod, jednocześnie w sensownym czasie i z rozsądnym użyciem "zasobów ludzkich" jakie mamy (a przez kilka ostatnich miesięcy nie było w zasadzie żadnych).

Reply:
Nie mówię o możliwie najlepszym kodzie - bo to jest kwestia zupełnie niezwiązana z moim postulatem. 
Chodzi o wydzielenie kodu tak by całość znajdowała się w pluginie. 
Żeby tak się stało, trzeba przeczytać dokumentację samego q2a, w której wszystko jest opisane:
https://docs.question2answer.org/plugins/modules/
https://docs.question2answer.org/plugins/modules-page/

Wystarczy pogrzebać. Nie jest to najbardziej dopracowana pod kątem UX dokumentacja, ale każdą zmianę można w prosty i szybki sposób zrobić tak, by nie ingerować w kod core. 
Nie ma problemu z tworzeniem nowych pages i przechwytywaniem z nich requestów, 
Nie ma problemu z przechwytywaniem requestów z istniejących pages. 
Można nawet sugerować strony. 

Można też nadpisać style bez używania qa-styles.css domyślnego motywu. 
W zasadzie dokumentacja pokazuje dokładnie w jaki sposób zrobić cokolwiek sie tylko chce - jest nawet 150 metod które można własnoręcznie nadpisać z core przy pomocy pluginu. 

Nie widzę żadnej potrzeby, konieczności ani trudności w tworzeniu zmian, które by nie wpływały na core. Jakość kodu w tym przypadku jest drugorzędna i jest rozwiązywana w trakcie pracy nad PR. 

Problemem nie jest samo tworzenie zmiany tylko nieczytanie dokładnie dokumentacji - co by masę rzeczy uprościło i przyspieszyło. Nie trzeba by robić haków na core, które zwiększają ryzyko wysypania się forum, którego nie da się przywrócić tylko wyłączeniem pluginu. Gdy zmieniamy coś w core to nie podziała wyłączenie pluginu tylko rollback zmian na kodzie który nie powinien być zmieniany - a niestety ma mase zmian, których nie powinien mieć. Szkoda pogłębiać tę przepaść. 

Idąc tym tokiem rozumowania - miesiąc poświęcony na aktualizację wersji może być czasem poświęconym bez sensu, bo wolałbyś, żebym zrobił 10 issues z tego projektu nawet jeśli miałyby być zrobione w core?

Reply:
Przeinaczasz na siłę to co chcę przekazać :) Zupełnie niepotrzebnie, bo ja naprawdę chciałbym jak najlepiej i poprawić to jak się da, a najlepiej cytując klasyka "zaorać" i zrobić od nowa. Tylko że się tak łatwo nie da. Chodzi mi o to, abyśmy nie utknęli na przenoszeniu prawie gotowych zmian na pluginy. Nie wiem na ile to jest trudne, bo nie zagłębiałem się w ten kod, do tego nie znam wszystkich możliwych przypadków i Ty też nie znasz. Dziwi mnie nieco więc, że wypowiadasz się z taką pewnością. Trochę rzeczy już w Q2A robiłem i nie było żadnego problemu z pluginami, z pewnymi miałem różnego rodzaju problemy. Być może problemy były z mojej winy, nie wiem, być może źle szukałem, co nie zmienia faktu że mogą się pojawić i może to potrwać, tym bardziej jak ktoś nie ma wprawy w dłubaniu w Q2A.

W związku z tym jeśli teraz znowu na miesiące zawiśniemy z prawie gotowymi zmianami, a może w ogóle nigdy nie wejdą jak komuś nie będzie chciało się zrobić/nie będzie wiedział jak, to tracimy coś, co moglibyśmy mieć. I oczywiście, najlepiej byłoby to zrobić jako plugin, tak, będzie super jak tak zrobimy. Ale nie ukrywajmy, że nie ma chętnych do pracy przy tym kodzie. Sam średnio miałem na to czas, chęci i możliwości, tak jak i Ty przez ostatnie miesiące nie interesowałeś się zupełnie tym co się tu działo. Cieszę się, że wróciłeś i coś się zaczyna dziać, wszyscy jakoś się zebrali i idzie, ale nie wiemy jak długo. W związku z tym wolę podchodzić tak, aby robić to na pluginach, ale jak się nie uda to próbować bez plugina. Próbować możliwie najlepiej, ale gdy się nie da to spróbować nieco gorzej, żeby mieć cokolwiek niż nie mieć nic. Bo nie ma kto tego niestety robić, nie mamy zatrudnionych ludzi.

> Nie widzę żadnej potrzeby, konieczności ani trudności w tworzeniu zmian, które by nie wpływały na core.

Cieszę się, że Ty nie widzisz, ale inni mogą mieć pewne trudności, sam nie raz je miałem.

> Idąc tym tokiem rozumowania - miesiąc poświęcony na aktualizację wersji może być czasem poświęconym bez sensu, bo wolałbyś, żebym zrobił 10 issues z tego projektu nawet jeśli miałyby być zrobione w core?

Zacytuję samego siebie:
> **Trzeba to wypośrodkować** - robić możliwie najlepszy kod, jednocześnie w sensownym czasie i z rozsądnym użyciem "zasobów ludzkich" jakie mamy

Aktualizacja silnika forum przynosi nowe funkcje, zmiany, optymalizacje, poprawki błędów - więc nie jest na pewno czasem poświęconym bez sensu. 10 issues wprowadziłoby pewnie wiele nowości i potrzebnych zmian, tak jak i aktualizacji forum wprowadzi wiele nowości i potrzebnych zmian.

To co podawałem to jak mówiłem przykłady, w tym konkretnym pull requeście nie wiem ile trzeba na przeniesienie tego na plugin, może w 15 minut autor to zrobi i cała gadka jest zbędna. Proponuję więc skończyć dyskusję, bo nic ona już chyba nie wniesie. Nie przekonasz mnie, że jedyną wartością ma być sposób/jakość pisanego kodu, a czas i koszt wykonania jest nieważny - to zawsze w jakimś stopniu jest ważne, trzeba znaleźć dobry środek. Moja propozycja: starajmy się wszystko robić na pluginach, ale jeśli się nie komuś nie uda, a dana funkcjonalność jest fajna/potrzebna i nie ma innego pomysłu jak to zrobić bez pluginu oraz innej osoby, która mogłaby to przejąć i zrobić jako plugin, miejmy z tyłu głowy pomyślenie o opcji dodania czegoś bez plugina. Ale tylko gdzieś z tyłu, domyślnie myślmy o wszystkim jako pluginie. Tylko tyle, proste.